netmaster/daemon.go

```diff
@@ -0,0 +1,339 @@
+
+// XXX: This function should be returning logical state instead of driver state
+func (d *daemon) endpoints(id string) ([]core.State, error) {
+	var (
+		err error
+		ep  *drivers.OvsOperEndpointState
+	)
+
+	ep = &drivers.OvsOperEndpointState{}
+	if ep.StateDriver, err = utils.GetStateDriver(); err != nil {
+		return nil, err
+	}
+
+	if id == "all" {
+		eps, err := ep.ReadAll()
+		if err != nil {
+			return []core.State{}, nil
+		}
+		return eps, nil
+	}
+
+	err = ep.Read(id)
+	if err == nil {
+		return []core.State{core.State(ep)}, nil
+	}
+
+	return nil, core.Errorf("Unexpected code path. Recieved error during read: %v", err)
+}
+
+// XXX: This function should be returning logical state instead of driver state
+func (d *daemon) networks(id string) ([]core.State, error) {
+	var (
+		err error
+		nw  *mastercfg.CfgNetworkState
+	)
+
+	nw = &mastercfg.CfgNetworkState{}
+	if nw.StateDriver, err = utils.GetStateDriver(); err != nil {
+		return nil, err
+	}
+
+	if id == "all" {
+		return nw.ReadAll()
+	} else if err := nw.Read(id); err == nil {
+		return []core.State{core.State(nw)}, nil
+	}
+
+	return nil, core.Errorf("Unexpected code path")
+}
+
+// runLeader runs leader loop
+func (d *daemon) runLeader() {
+	router := mux.NewRouter()
+
+	// acquire listener mutex
+	d.listenerMutex.Lock()
+	defer d.listenerMutex.Unlock()
+
+	// Create a new api controller
+	d.apiController = objApi.NewAPIController(router)
+
+	// initialize policy manager
+	mastercfg.InitPolicyMgr(d.stateDriver, d.ofnetMaster)
+
+	// Register netmaster service
+	d.registerService()
+
+	// setup HTTP routes
+	d.registerRoutes(router)
+
+	// Create HTTP server and listener
+	server := &http.Server{Handler: router}
+	listener, err := net.Listen("tcp", d.listenURL)
+	if nil != err {
+		log.Fatalln(err)
+	}
+
+	log.Infof("Netmaster listening on %s", d.listenURL)
+
+	// start server
+	go server.Serve(listener)
+
+	// Wait till we are asked to stop
+	<-d.stopLeaderChan
+
+	// Close the listener and exit
+	listener.Close()
+	log.Infof("Exiting Leader mode")
+}
+
+// runFollower runs the follower FSM loop
+func (d *daemon) runFollower() {
+	router := mux.NewRouter()
+	router.PathPrefix("/").HandlerFunc(slaveProxyHandler)
+
+	// acquire listener mutex
+	d.listenerMutex.Lock()
+	defer d.listenerMutex.Unlock()
+
+	// start server
+	server := &http.Server{Handler: router}
+	listener, err := net.Listen("tcp", d.listenURL)
+	if nil != err {
+		log.Fatalln(err)
+	}
+
+	// start server
+	go server.Serve(listener)
+
+	// Register netmaster service
+	d.registerService()
+
+	// just wait on stop channel
+	log.Infof("Listening in follower mode")
+	<-d.stopFollowerChan
+
+	// Close the listener and exit
+	listener.Close()
+	log.Info("Exiting follower mode")
+}
+
+// becomeLeader changes daemon FSM state to master
+func (d *daemon) becomeLeader() {
+	// ask listener to stop
+	d.stopFollowerChan <- true
+
+	// set current state
+	d.currState = "leader"
+
+	// Run the HTTP listener
+	go d.runLeader()
```

Comment:
It can happen while transitioning from follower to leader. We issue a stop follower before becoming leader, but the actual stop happens in another thread.